frontend/src/scenes/surveys/surveyActivityDescriber.tsx

```diff
@@ -410,3 +440,177 @@ function getPreposition(field: string): string {
             return 'of'
     }
 }
+
+type SurveyQuestion = BasicSurveyQuestion | LinkSurveyQuestion | RatingSurveyQuestion | MultipleSurveyQuestion
+
+export function describeQuestionChanges(before: SurveyQuestion, after: SurveyQuestion): JSX.Element[] {
+    const commonChanges = describeCommonChanges(before, after)
+    const typeChangeDescription =
+        before.type !== after.type
+            ? [
+                  <>
+                      changed question type from <strong>{before.type}</strong> to <strong>{after.type}</strong>
+                  </>,
+              ]
+            : []
+
+    const specificChanges = match([before, after] as const)
```

Comment:
some cool shit in here with `ts-pattern` that lets me match a tuple of types with a very nice combinator-style developer experience.